src/deluge/gui/views/instrument_clip_view.cpp

```diff
@@ -1282,93 +1282,112 @@ const uint32_t auditionPadActionUIModes[] = {UI_MODE_AUDITIONING,
+						continue;
+					}
+
+					char const* currentPathChars = afh->filePath.get();
+
+					if (currentPathChars == NULL) {
+						continue;
+					}
+					char const* slashAddress = strrchr(currentPathChars, '/');
+					if (slashAddress) {
+						int32_t slashPos = (uint32_t)slashAddress - (uint32_t)currentPathChars;
+						String dir;
+						dir.set(&afh->filePath);
+						dir.shorten(slashPos);
+						FRESULT result = f_opendir(&staticDIR, dir.get());
+						if (result != FR_OK) {
+							display->displayError(ERROR_SD_CARD);
+							return ActionResult::DEALT_WITH;
 						}
-						audioFileManager.loadAnyEnqueuedClusters();
-						fnArray[numSamples].set(staticFNO.fname);
-						numSamples++;
-						if (numSamples >= MaxFiles) {
-							break;
+
+
+						int32_t count = 0;
+						while (true) {
+							audioFileManager.loadAnyEnqueuedClusters();
+							result = f_readdir(&staticDIR, &staticFNO);	 /* Read a directory item */
+							if (result != FR_OK || staticFNO.fname[0] == 0) {
+								break; // Break on error or end of dir
+							}
+							if (staticFNO.fname[0] == '.' || staticFNO.fattrib & AM_DIR || !isAudioFilename(staticFNO.fname) ) {
+								continue; // Ignore dot entry
+							}
+							if(random(count++)==0) {	// Algorithm: Reservoir Sampling with k=1
+								chosen_fn.set(staticFNO.fname);
+							}
 						}
-					}
 
-					if (numSamples >= 2) {
-						soundDrum->unassignAllVoices();
-						afh->setAudioFile(NULL);
-						String filePath; //add slash
-						filePath.set(&dir);
-						int32_t dirWithSlashLength = filePath.getLength();
-						if (dirWithSlashLength) {
-							filePath.concatenateAtPos("/", dirWithSlashLength);
-							dirWithSlashLength++;
+						if (count) {
+							char const* fn = chosen_fn.get();
+							soundDrum->unassignAllVoices();
+							afh->setAudioFile(NULL);
+							String filePath; //add slash
+							filePath.set(&dir);
+							int32_t dirWithSlashLength = filePath.getLength();
+							if (dirWithSlashLength) {
+								filePath.concatenateAtPos("/", dirWithSlashLength);
+								dirWithSlashLength++;
+							}
+							filePath.concatenateAtPos(fn, dirWithSlashLength);
+							AudioEngine::stopAnyPreviewing();
+							afh->filePath.set(&filePath);
+							afh->loadFile(false, true, true, 1, 0, false);
+							soundDrum->name.set(fn);
+							((Instrument*)currentSong->currentClip->output)->beenEdited();
 						}
-						char const* fn = fnArray[random(numSamples - 1)].get();
-						filePath.concatenateAtPos(fn, dirWithSlashLength);
-						AudioEngine::stopAnyPreviewing();
-						afh->filePath.set(&filePath);
-						afh->loadFile(false, true, true, 1, 0, false);
-						soundDrum->name.set(fn);
-						numRandomized++;
-						getCurrentInstrument()->beenEdited();
 					}
 				}
-			}
-		}
-		if (numRandomized > 0) {
-			display->displayPopup(deluge::l10n::get(deluge::l10n::String::STRING_FOR_RANDOMIZED));
-			return ActionResult::DEALT_WITH;
 		}
+		display->displayPopup(chosen_fn.get());
```

Comment:
I like the idea of showing the file name here, but have a few UI concerns:
- Does this work with filenames longer than what can be shown across the OLED screen? I'm not sure we handle wide popups like that gracefully
- This is mostly meaningless for the randomize all case

Reply:
a) Fair point, but the file name is currently already shown when pressing the audition pad, so if it works there...
And if it does not, the problem should be addressed in the popup function...

b) Agreed, but IMO still more informative than "Randomized". But if you prefer "Randomized all" or something else, I am fine with it.

Reply:
Let's use "randomized all" for the all case.